mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -1,149 +1,505 @@
-        max_conf: u32,
-        watch_only: bool,
-        addresses: &[&str],
-    ) -> UtxoRpcFut<Vec<ZUnspent>> {
-        let fut = rpc_func!(self, "z_listunspent", min_conf, max_conf, watch_only, addresses);
-        Box::new(fut.map_to_mm_fut(UtxoRpcError::from))
+    #[inline]
+    pub(super) fn current_block(&self) -> BlockHeight { self.sync_handle.as_ref().expect("always Some").current_block }
+}
+
+pub enum SyncStatus {
+    UpdatingBlocksCache {
+        current_scanned_block: u64,
+        latest_block: u64,
+    },
+    BuildingWalletDb,
+    Finished {
+        block_number: u64,
+    },
+}
+
+pub struct SaplingSyncLoopHandle {
+    pub(super) current_block: BlockHeight,
+    grpc_client: CompactTxStreamerClient<Channel>,
+    db: WalletDbShared,
+    consensus_params: ZcoinConsensusParams,
+    /// Notifies about sync status without stopping the loop, e.g. on coin activation
+    sync_status_notifier: AsyncSender<SyncStatus>,
+    /// If new tx is required to be generated, we stop the sync and respawn it after tx is sent
+    /// This watcher waits for such notification
+    on_tx_gen_watcher: AsyncReceiver<OneshotSender<Self>>,
+    pub(super) watch_for_tx: Option<TxId>,
+}
+
+impl SaplingSyncLoopHandle {
+    fn notify_blocks_cache_status(&mut self, current_scanned_block: u64, latest_block: u64) {
+        if self
+            .sync_status_notifier
+            .try_send(SyncStatus::UpdatingBlocksCache {
+                current_scanned_block,
+                latest_block,
+            })
+            .is_err()
+        {
+            debug!("No one seems interested in SyncStatus");
+        }
+    }
+
+    fn notify_building_wallet_db(&mut self) {
+        if self
+            .sync_status_notifier
+            .try_send(SyncStatus::BuildingWalletDb)
+            .is_err()
+        {
+            debug!("No one seems interested in SyncStatus");
+        }
     }
 
-    fn z_send_many(&self, from_address: &str, send_to: Vec<ZSendManyItem>) -> UtxoRpcFut<String> {
-        let fut = rpc_func!(self, "z_sendmany", from_address, send_to);
-        Box::new(fut.map_to_mm_fut(UtxoRpcError::from))
+    fn notify_sync_finished(&mut self) {
+        if self
+            .sync_status_notifier
+            .try_send(SyncStatus::Finished {
+                block_number: self.current_block.into(),
+            })
+            .is_err()
+        {
+            debug!("No one seems interested in SyncStatus");
+        }
     }
 
-    fn z_import_key(&self, key: &str) -> UtxoRpcFut<()> {
-        let fut = rpc_func!(self, "z_importkey", key, "no");
-        Box::new(fut.map_to_mm_fut(UtxoRpcError::from))
+    async fn update_blocks_cache(&mut self) -> Result<(), MmError<UpdateBlocksCacheErr>> {
+        let request = tonic::Request::new(ChainSpec {});
+        let current_blockchain_block = self.grpc_client.get_latest_block(request).await?;
+        let current_block_in_db = block_in_place(|| self.db.lock().blocks_db.get_latest_block())?;
```

Comment:
https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html
> Be aware that although this function avoids starving other independently spawned tasks, any other code running concurrently in the same task will be suspended during the call to block_in_place. This can happen e.g. when using the [join!](https://docs.rs/tokio/latest/tokio/macro.join.html) macro. To avoid this issue, use [spawn_blocking](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) instead of block_in_place.

It's used in the `light_wallet_db_sync_loop`, which is a standalone spawned task that is not joined/selected over. We can safely use `block_in_place` here.